Document the Fluent Plugin (#228)
grammyjs/website: site/docs/plugins/i18n-fluent.md

# Internationalization with Fluent

> [Fluent](https://projectfluent.org/) is a localization system for natural-sounding translations created and maintained by Mozilla Foundation.

Reviewer: ```suggestion
[Fluent](https://projectfluent.org/) is a localization system made by the Mozilla Foundation for natural-sounding translations.
```

> It has a very powerful and at the same time elegant syntax, that enables even non-developer people to efficiently work with translation files.

Reviewer: ```suggestion
It has a very powerful and elegant syntax that lets anyone write efficient and fully-understandable translations.
```

> Thanks to the [grammY Fluent integration plugin](https://github.com/grammyjs/fluent) you can easily translate your bot to multiple languages using industry best practices like variable substitution, pluralization, locale negotiation, conditional messages and message reuse.

Reviewer: ```suggestion
This plugin takes advantage of this amazing localization system to make grammY-powered bots fluent with high-quality translations.
```

## Install

Install the dependencies:

```shell
npm install --save @grammyjs/fluent @moebius/fluent
```

grammY Fluent is built on top of the `@moebius/fluent` package, so you will need to install it as well.
Also, it is highly advisable to read the documentation of the [original library](https://github.com/the-moebius/fluent#moebiusfluent).

## Instantiate Fluent

The next step would be to create a Fluent instance:

```typescript
import { Fluent } from '@moebius/fluent';

const fluent = new Fluent();
```

Then, you will need to add at least one translation to the Fluent instance:

```typescript
await fluent.addTranslation({
  
  // Specify locale supported by your translation,
  // you can specify multiple locales if you want
  locales: 'en',
  
  // You can specufy the translation content directly:
  source: '{YOUR TRANSLATION FILE CONTENT}',

  // OR you can specify a path to translation file or multiple files:
  filePath: [
    `${__dirname}/feature-1/translation.en.ftl`,

>     `${__dirname}/feature-2/translation.en.ftl`

Reviewer 2: Does this work cross-platform?

Reviewer 2: To be clear, I am referring to file paths on Windows. I do not mean Deno & Node.

Author: All the examples are meant for Node.js and Linux. I don't think that we should complicate the example code to support Windows. The Windows devs should know what to do with this.

import { FluentContextFlavor } from '@grammyjs/fluent';

// Extend your application context type with the provided
// flavor interface
export type MyAppContext = (
  & Context
  & FluentContextFlavor
);
```

You will need to create you bot instance the following way in order to use the augmented context type:

```typescript
const bot = new Bot<MyAppContext>();
```

And the final step would be to register the Fluent plugin itself with grammY:

```typescript
// Add fluent middleware to the bot
bot.use(useFluent({
  fluent,
}));
```

Just make sure to pass the [previously created](#instantiate-fluent) Fluent instance.

## Render the Localized Messages

Great, now we have everything in place to render our translated message!
Let's do just that by defining a test command in our bot:

```typescript
bot.command('i18n_test', async context => {

  // Call the "translate" or "t" helper to render the
  // message by specifying it's ID and
  // additional parameters:
  await context.reply(context.t('welcome', {
    name: context.from.first_name,
    applesCount: 1,
  }));

});
```

Now, you can start your bot and use the `/i18n_test` command.
It should render the following message:

> ```text

Reviewer 2: ```suggestion
```text:no-line-numbers
```
